### PR TITLE
feat(apache_metrics source): Initial implementation

### LINKED ISSUE #3092: New `apache_metrics` source
For collecting Apache web server metrics.

### PR DESCRIPTION
**Pull request by jszwedko**
**EDIT** This is ready for review now. See comments below.

Closes #3092 

I am just pushing this up for any preliminary feedback before I get back to it on Tuesday. I'm definitely still on the upclimb of the Rust learning curve so feedback on Rust idioms and patterns especially welcome.

This is roughly based on the implementation of the `prometheus` source which operates in a similar manner: scraping an HTTP endpoint and parsing the payload to generate metrics.

I left some notes on where I think the architecture could be improved, below, but I think it'd be best to hold off until https://github.com/timberio/vector/issues/3702 to do a larger refactor of how this and the `prometheus` source work.

The namespace handling will likely change as a part of #3686 

TODOs:

- [x] Tests in `apache_metrics/mod.rs`
- [x] Testing parsing failures
- [x] Applying `host` and `endpoint` tags
- [x] Handling request errors and publishing `apache_up=0`
- [x] Documentation

### CODE REVIEW

File: src/sources/apache_metrics/mod.rs
Comment (jszwedko) [inline src/sources/apache_metrics/mod.rs:86]: I'd really like to break up the below into separate functions but figured I'd wait for #3702 . This largely matches the `prometheus` source in its current form.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [inline src/sources/apache_metrics/parser.rs:24]: I'm curious to get feedback or thoughts on how to model this better where the parse function wants to return metrics _and_ errors rather than simply one or the other. This was a design choice I made to still publish metrics that we could parse even if there was an error with one of them.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [inline src/sources/apache_metrics/parser.rs:78]: Is this sort of `enum` idiomatic?

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [inline src/sources/apache_metrics/parser.rs:600]: I'll update with code comments, but the difference between this and the above test is that the output is different depending on whether `ExtendedStatus` is enabled or not.

https://httpd.apache.org/docs/2.4/mod/core.html#extendedstatus

File: src/sources/apache_metrics/mod.rs
Comment (jszwedko) [inline src/sources/apache_metrics/mod.rs:89]: I'd like to get rid of this clone, but was unable to figure out how without annoying the borrow checker.

File: src/sources/apache_metrics/mod.rs
Comment (jszwedko) [inline src/sources/apache_metrics/mod.rs:106]: Same here.

File: src/sources/apache_metrics/mod.rs
Comment (jszwedko) [inline src/sources/apache_metrics/mod.rs:197]: These tests just test high-level behavior; there are separate tests in `apache_metrics/parser.rs` that test metric generation behavior.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [inline src/sources/apache_metrics/parser.rs:40]: Thoughts here? I did this to dedupe the lines which required sorting.

**Review by jszwedko — COMMENTED** [review]
I think this is ready for review now!

Again, thoughts on patterns and idioms especially welcome. I tried to call out a couple of places where I had a question. I'll drop the `TODOs` from the code before merging this after I see people's thoughts on them.

**Review by jamtur01 — COMMENTED** [review]
Core stuff looks good to me. I am skeptical I can add value reviewing any of the actual code. :)

File: src/sources/apache_metrics/mod.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/mod.rs:195]: I don't believe this is needed, since the whole module is behind the same feature flag.

File: .github/semantic.yml
Comment (JeanMertz) [inline .github/semantic.yml:105]: Was this auto-generated from `.github/semantic.yml.erb`? (and if so, how did you do that 👀).

I need to add one for the `remap` transform as well, but I haven't looked into where the actual `.metadata` data comes from when it generates this file.

File: src/sources/apache_metrics/mod.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/mod.rs:1]: nit: since this module consists of a single file, you can put it in `src/sources/apache_metrics.rs`.

Basically, these two are exactly the same:

* `src/sources/apache_metrics.rs`
* `src/sources/apache_metrics/mod.rs`

The advantage of the second one is if you have multiple modules that you want to namespace under the `apache_metrics` namespace (and potentially re-export the most common public interface at that level).

Comment (JeanMertz) [reply src/sources/apache_metrics/mod.rs:1]: Whoops, never mind, there's actually a `parser.rs` file that was collapsed in the GH UI, so I missed it.

All good 😄

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [reply src/sources/apache_metrics/parser.rs:24]: I wonder if an iterator-based approach would work well?

The iterator would return `Option<Result<Metric, ParseError>>`. The caller can continue iterating, even if an error is returned, until no more metrics are available.

You can't have it return a metric _and_ error simultaneously, but I don't believe that's needed?

Regardless, what you have here works nicely, so I wouldn't worry too much about changing this.

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [reply src/sources/apache_metrics/parser.rs:40]: I'm not quite sure what the purpose of this de-duplication is.

Is it our job to filter out those duplicate status events, if it's what Apache is reporting? What happens if I call this function twice, it would filter out duplicates for each individual call, but I would still end up with two similar events since it doesn't filter duplicates across calls, right?

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [reply src/sources/apache_metrics/parser.rs:78]: This seems similar to `Option<Result<Vec<Metric>, ParseError>>`?

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/parser.rs:92]: Perhaps we could move these keys into a (documented) enum?

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/parser.rs:146]: I see a reasonable amount of duplication in the next couple of match arms.

Perhaps use a single arm, and then split up where needed? e.g.

```rust
"CPUUser " | "CPUSystem" | /* ... */ => {
	// ...

	match { /* ... */ }
}
```

Can't say for certain how much clean-up that would provide, but might be worth a try.

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/parser.rs:363]: Seems like we should be able to move this into a constant to avoid the allocations on each cycle?

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/parser.rs:368]: `unwrap_or_default()` should work here as well I believe.

Comment (JeanMertz) [reply src/sources/apache_metrics/parser.rs:368]: Although that would probably require something similar to `(*scores.get(c)).unwrap_or_default()` to dereference the `u32`.

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [inline src/sources/apache_metrics/parser.rs:377]: Maybe swap the cases, to avoid the double negative?

**Review by JeanMertz — APPROVED** [review]
Sweet! Left a couple of comments, but didn't dive too deep into the Apache metrics logic itself. I can circle back later, but let's wait if someone with more knowledge comes along to review 😄

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:40]: I'll try to improve the comment there, but basically there is a bug in mod_status where, in the text output, `IdleWorkers` and `BusyWorkers` appear twice. If I don't dedupe we end up publishing duplicate metrics, which is probably ok, but seems a little wonky.

Apache bug: https://bz.apache.org/bugzilla/show_bug.cgi?id=63300

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:78]: :thinking: yeah it is basically that. I'll try swapping to remove the type.

File: .github/semantic.yml
Comment (jszwedko) [reply .github/semantic.yml:105]: I actually missed that this was a generated file. I'm actually not sure how to regenerate it. I assume @binarylogic knows? Does this repo need a `generate.rb` like `vector-website` has?

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:24]: Hmm, yeah, I think I like that better. I'll give it a shot.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:92]: Agree :+1:

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:146]: It is a bit duplicative, but I actually like the simplicity of just mapping each string rather than parsing the key names. If you feel strongly though, I'll give it a shot.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:363]: Agreed :+1:

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:368]: :+1: I ended up with `scores.get(c).copied().unwrap_or_default()`.

**Comment by jszwedko** [comment]
@JeanMertz thanks for the thorough review and recommendations! I think I implemented them all in https://github.com/timberio/vector/pull/3704/commits/5e06a5273c512210b88d8ef1910949e28fa6a91d if you want to take a look. I think it is cleaner now.

**Comment by jszwedko** [comment]
**Note** I'll still drop the TODOs I had in the code before merging this; I just put them there to call out questions I had.

File: src/sources/apache_metrics/parser.rs
Comment (JeanMertz) [reply src/sources/apache_metrics/parser.rs:198]: Ah, right 👍

I guess one way to avoid having to collect is to keep track of the field through a boolean, and return `None` in the `filter_map` above as soon as we've flipped the boolean to `true`?

Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs]: :+1: This made me realize I could also just use a `HashMap` which would take care of making it unique: 8e3d7539 . I'm curious what you think of that.

File: src/sources/apache_metrics/parser.rs
Comment (bruceg) [inline src/sources/apache_metrics/parser.rs:436]:
```suggestion
    match namespace {
        "" => name.to_string(),
        _ => format!("{}_{}", namespace, name),
    }
```

File: src/sources/apache_metrics/parser.rs
Comment (bruceg) [inline src/sources/apache_metrics/parser.rs:215]: Can this have a concrete type, or are there multiple types of error handled here? If multiple types, did you consider an `enum ParseError`?

File: src/sources/apache_metrics/parser.rs
Comment (bruceg) [inline src/sources/apache_metrics/parser.rs:236]: Could this return an iterator in the place of the `Vec`? Everything but the `Scoreboard` has a single value and could use `std::iter::once`.

File: src/sources/apache_metrics/parser.rs
Comment (bruceg) [inline src/sources/apache_metrics/parser.rs:245]: This looks like there would be some opportunity for refactoring the duplication, but I see there is a few different things going on with `tags` and `value`.

**Review by bruceg — COMMENTED** [review]
Looks pretty good from an idiomatic Rust point of view. There's definite room for refactoring, but I understand you want to hold off on some parts of that.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:215]: I believe there are at least two types of parse errors, one for int and one for float. I think I did, at one point, have it as an enum, but it didn't seem worth it since all we were doing was just propagating it up without handling.

I guess, generally, do you feel it's better to be specific with the error types even if the caller doesn't care?

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:31]: I thought about that, but figured since they are just passed through directly to the `Metric` type which wants a `f64`s anyway, that converting back and forth didn't really give as much. However, I can see the argument for modeling the data as it is. I'll make that change.

File: src/sources/apache_metrics/parser.rs
Comment (bruceg) [reply src/sources/apache_metrics/parser.rs:215]: Generally, I think it's better to be specific with _all_ types, even if the caller doesn't care, as it avoids invoking more "mechanism" at run time (allocations, dynamic dispatch, etc). However, it doesn't matter as much when it isn't the hot path and, by definition, exceptional errors are not the hot path. In the case of errors, though, adding a layer of wrapping adds context as to why the error happened. So it's partly a judgment call.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs:59]: This feels cleaner, thanks for the suggestion!

The compiler did walk me into quite the type signature:

```rust
fn parse_value<T: std::str::FromStr>(key: &str, value: &str) -> Result<T, ParseError>                     
where                                                                                                     
    <T as std::str::FromStr>::Err: std::error::Error,                                                     
    <T as std::str::FromStr>::Err: 'static,                                                               
{                                                                                                         
    value.parse::<T>().map_err(|err| ParseError {                                                         
        key: key.to_string(),                                                                             
        err: err.into(),                                                                                  
    })                                                                                                    
}  
```

Comment (bruceg) [reply src/sources/apache_metrics/parser.rs:59]: Even though the compiler recommends it, I think you can simplify the `where` clause since you already have `T: FromStr`:
```rust
fn parse_value<T>(key: &str, value: &str) -> Result<T, ParseError>
where
    T: FromStr,
    T::Err: std::error::Error + 'static
```

**Comment by jszwedko** [comment]
@bruceg thanks for all of the thorough feedback! It was helpful. I think I implemented all of the suggestions aside from the general refactoring one. I tried to keep the commits atomic so you could just review them in order from the time of your review.

For the general refactoring comment (https://github.com/timberio/vector/pull/3704#discussion_r488839729), I definitely agree that it could use some love. I still feel like it might be better to wait for #3702 to spend a lot of time on it, but, if you feel strongly about refactoring some sooner, would you be up for pairing on it? I think that'd help me work through it a bit more quickly. Maybe we could sketch out what #3702 might look like too.

File: src/sources/apache_metrics/parser.rs
Comment (jszwedko) [reply src/sources/apache_metrics/parser.rs]: :+1: that makes sense. It seems more flexible too in that if we add another caller that would care, the specificity would already be there. Thanks for the explanation.

File: src/sources/apache_metrics/parser.rs
Comment (bruceg) [inline src/sources/apache_metrics/parser.rs]: These can combine, and I'm not sure the `as FromStr` is necessary:
```suggestion
    T::Err: std::error::Error + 'static,
```

**Review by bruceg — APPROVED** [review]
I'm kinda sorry I suggested looking into this path. This turned a relatively simple error struct into something pretty verbose, all to avoid a boxed trait. I had forgotten you would be working with both `ParseFloatError` and `ParseIntError`.

**Comment by jszwedko** [comment]
> I'm kinda sorry I suggested looking into this path. This turned a relatively simple error struct into something pretty verbose, all to avoid a boxed trait. I had forgotten you would be working with both ParseFloatError and ParseIntError.

It's definitely more verbose, but I do like that there is a tighter contract on the error types. It feels like this would be a common pattern for types that can return multiple types of errors?

Thanks again for the reviews!

**Comment by bruceg** [comment]
> It feels like this would be a common pattern for types that can return multiple types of errors?

Definitely, and that is what our use of the snafu crate is designed to help with. It might (have) helped with this, though your usage here is a little different than other places.

**Comment by jszwedko** [comment]
Opened https://github.com/timberio/vector/issues/4003 for the flakey mac failure.

The shutdown linux failure (`remap_function_now`) doesn't seem to be flakey, but is on master already. I pinged Jean about it since he was working in that area.

**Comment by jamtur01** [comment]
Boo ya! 🥳 🎆 🍾 

Awesome stuff.

File: src/sources/apache_metrics/mod.rs
Comment (fanatid) [inline src/sources/apache_metrics/mod.rs:106]: From [rfc/2020-08-21-3092-apache-metrics-source.md](https://github.com/timberio/vector/blob/3d11ab33f6a3ae3de6e2b1a73f97ca23879720a8/rfcs/2020-08-21-3092-apache-metrics-source.md):
> `endpoint` the full endpoint (sans any basic auth credentials)

Should we remove credentials here?

Comment (jszwedko) [reply src/sources/apache_metrics/mod.rs:106]: Yes we should :smile: Thanks for catching that!

Comment (jszwedko) [reply src/sources/apache_metrics/mod.rs:106]: I started to fix this, but will have to pick it back up later. It wasn't as straightforward as I thought it might be.

Comment (jszwedko) [reply src/sources/apache_metrics/mod.rs:106]: https://github.com/timberio/vector/pull/4410

**Pull request merged**
